File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/addressspace/nodemanager/NodeManagerBase.java
package org.opcfoundation.webserver.addressspace.nodemanager;

import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;
import org.eclipse.milo.opcua.stack.core.types.builtin.StatusCode;
import org.eclipse.milo.opcua.stack.core.types.structured.*;
import org.opcfoundation.webapi.service.types.*;
import org.opcfoundation.webserver.service.transactions.base.*;
import org.springframework.lang.Nullable;
import org.opcfoundation.webserver.addressspace.nodes.UaNode;
import org.opcfoundation.webserver.types.common.UaBrowseAdditionalInfo;

import java.util.ArrayList;
import java.util.List;

public interface NodeManagerBase {
    int nsIndex();

    String namespaceUri();

    @Nullable UaNode getNode(NodeId nodeId);

    default void onStartUp() throws UaRuntimeException {}
    default void onShutDown() throws UaRuntimeException {}

    default UaBrowseTransaction getBrowseTransaction(
            ServiceContext context,
            BrowseDescription nodeToBrowse,
            UaBrowseAdditionalInfo additionalInfo,
            int handleId)
    {
        UaBrowseTransaction ret = new UaBrowseTransaction(
                context,
                nodeToBrowse,
                additionalInfo,
                handleId);

        ret.setStatusCode(StatusCode.of(StatusCodes.Bad_NodeIdUnknown));
        return ret;
    }

    default List<UaReadTransaction> getReadTransactions(
            ReadContext context,
            List<Integer> handleIds)
    {
        UaReadTransaction transaction = new UaReadTransaction(
                context,
                handleIds);

        ArrayList<UaReadTransaction> transactions = new ArrayList<>();
        transactions.add(transaction);

        return transactions;
    }

    default List<UaWriteTransaction> getWriteTransactions(
            WriteContext context,
            List<Integer> handleIds)
    {
        UaWriteTransaction transaction = new UaWriteTransaction(
                context,
                handleIds);

        ArrayList<UaWriteTransaction> transactions = new ArrayList<>();
        transactions.add(transaction);

        return transactions;
    }

    default UaMethodCallTransaction getMethodCallTransaction(
            CallContext context,
            int handleId)
    {
        return new UaMethodCallTransaction(
               context,
               handleId);
    }

    default UaHistoryReadTransaction getHistoryReadTransaction(
            HistoryReadContext context,
            int handleId)
    {
        return new UaHistoryReadTransaction(
                context,
                handleId);
    }
}

File: opcua-webservice-node/src/server/addressspace/nodemanager/NodeManagerBase.ts
import { StatusCodes } from "opcua-webapi";
import { makeUaStatusCode, UaBrowseDescription, UaNodeId } from "opcua-webapi-ts";
import { UaNode } from "../nodes";
import { CallContext, HistoryReadContext, ReadContext, ServiceContext, UaBrowseAdditionalInfo, WriteContext } from "../../types";
import { UaBrowseTransaction, UaHistoryReadTransaction, UaMethodCallTransaction, UaReadTransaction, UaWriteTransaction } from "../../service/transactions";

export abstract class NodeManagerBase {
        abstract nsIndex(): number;

        abstract namespaceUri(): string;

        abstract getNode(nodeId: UaNodeId): UaNode | null;

        async onStartUp(): Promise<void> {
                return;
        }

        async onShutDown(): Promise<void> {
                return;
        }

        getBrowseTransaction(
                context: ServiceContext,
                nodeToBrowse: UaBrowseDescription,
                additionalInfo: UaBrowseAdditionalInfo,
                handleId: number): UaBrowseTransaction {

                const transaction = new UaBrowseTransaction(
                                context,
                                nodeToBrowse,
                                additionalInfo,
                                handleId);

                transaction.setStatusCode(makeUaStatusCode(StatusCodes.BadNodeIdUnknown));
                return transaction;
        }

        getReadTransactions(
                context: ReadContext,
                handleIds: Array<number>): Array<UaReadTransaction> {
                return [new UaReadTransaction(context, handleIds)];
        }

        getWriteTransactions(
                context: WriteContext,
                handleIds: Array<number>): Array<UaWriteTransaction> {
                return [new UaWriteTransaction(context, handleIds)];
        }

        getMethodCallTransaction(
                context: CallContext,
                handleId: number): UaMethodCallTransaction {
                return new UaMethodCallTransaction(context, handleId);
        }

        getHistoryReadTransaction(
                context: HistoryReadContext,
                handleId: number): UaHistoryReadTransaction {
                return new UaHistoryReadTransaction(context, handleId);
        }
}